### ISSUE TITLE
Add a mechanism for avoiding package breakage when Base adds new names in a minor release

### ISSUE THREAD

**Issue by CameronBieganek** [JuliaLang/julia#52540]
This subject has come up before, most recently in #52535, but I am not aware of an issue that tracks this.

When Base adds new names in a minor release, packages that were already exporting those names become broken. Here's a MWE where the thought experiment is that A exports a function `length`, and then later Base adds and exports a function `length`:

```julia
julia> module A
           length(x) = 42
           export length
       end
Main.A

julia> using .A

julia> length([1, 2])
WARNING: both A and Base export "length"; uses of it in module Main must be qualified
ERROR: UndefVarError: `length` not defined
Stacktrace:
 [1] top-level scope
   @ REPL[3]:1
```

When Package A depends on package B, it can protect itself from breakage due to new names in B by only importing the names it uses from B, like `using B: foo, bar`. However, there is not such mechanism for Base, since there is an implicit `using Base` at the beginning of a REPL session.

We need some mechanism for preventing package breakage when Base adds a new exported name in a minor release. I'm sure I'll get blowback for this, but as it stands right now one of the following two things must happen in order for Julia and its ecosystem to truly follow semantic versioning:

1. Base Julia should stop adding names in minor releases, because it is a breaking change for any package that already exports that name.
2. Packages should upper bound the version of Julia, because any minor release of Julia could break their package.

Don't shoot the messenger. I'm just enumerating what the rules of semantic versioning require. 😅

**Comment by jariji** [comment]
The error occurred in your code because the user didn't import the names they wanted from `A` as they should have to guarantee that their code would keep working after Base introduced `length`.

**Comment by CameronBieganek** [comment]
There is no such requirement on users. Users are not required to follow semantic versioning. It is allowed for users to write a script like this:

```julia
using A
length([1, 2])
```

Suppose we are using version 1.0 of package A, and the compat section of the Project.toml file for version 1.0 of package A contains this entry:

```
julia = "1.6"
```

Then package A is declaring that version 1.0 of package A will work for all versions of Julia greater than or equal to 1.6 and less than 2.0. But suppose Julia 1.7 introduces and exports `length`. Then version 1.0 of package A will be broken on Julia 1.7, even though the compat section declares that it should work.

In other words, this unit test for version 1.0 of package A will pass on (the hypothetical) Julia 1.6 and fail on Julia 1.7:

```julia
using A
using Test
@test length([1, 2]) == 42
```

To reiterate, end user code is under no requirement to use semantic versioning. It is 100% allowed to write `using A` and expect that the code will work if the package manager has installed a compatible set of packages.

> See https://github.com/JuliaLang/julia/pull/42080.

Thanks---I've read that issue many times already. It does not address the issue with Base adding new names.

**Comment by CameronBieganek** [comment]
This is really the key point: If the package manager has installed a compatible set of packages, then `using A; foo()` should always work. Updating to a declared compatible Julia release while keeping the version of A fixed should not break `using A; foo()`.

**Comment by jariji** [comment]
I think the situation is that there is a restriction on the conditions under which Julia code is guaranteed to keep working. But that restriction is afaik undocumented and should become documented in some change like https://github.com/JuliaLang/julia/pull/42080.

Under that restriction, a user can add a third option to your list: list the names she imports if she wants to ensure compatibility through upgrades.

**Comment by CameronBieganek** [comment]
Yeah, the title of #42080 is "recommend against `using Foo` in package code", but if the recommendation was extended to scripts, that would protect script code from breakage. However, I think scripts like the following are pervasive:

```julia
using A, B, C
foo(); bar(); baz()
```

Declaring that those scripts are no longer expected to work across minor versions of Julia almost feels like a breaking change to the language...

**Comment by CameronBieganek** [comment]
@LilithHafner made this suggestion in a [comment](https://github.com/JuliaLang/julia/pull/42080#issuecomment-1745706532) on the "recommend against `using Foo` PR":

> Package-Base conflicts should IMO be resolved by resolving ambiguities in favor of non-base exports, possibly with a runtime warning.

Package-Base conflicts currently throw an error, so, if I understand correctly, resolving them in favor of non-Base exports could be done in a Julia 1.X release.

**Comment by jariji** [comment]
Python has that feature but I don't think it's a great idea. Implicitly choosing a function can turn exceptions into incorrect behavior. 

For [example](https://twitter.com/jakevdp/status/568211180642177026):

```py
>>> start = -1
>>> sum(range(5), start)
9
>>> from numpy import *
>>> sum(range(5), start)
10
```
because the second argument of `builtins.sum` is the starting value but the second argument of `numpy.sum` is the axis to sum over.


This can happen within a release or across releases, if package `P` publishes a name that Base is using but a user script contains `using P` then the meaning will switch.
